pkg/extension/sumologicextension/extension.go

```diff
@@ -148,43 +148,86 @@ func (se *SumologicExtension) validateCredenials(
 
 func (se *SumologicExtension) Start(ctx context.Context, host component.Host) error {
 	se.logger.Info(banner)
-	colCreds, registrationDone, err := se.getCredentials(ctx)
+
+	colCreds, registrationDone, err := se.handleCredentials(ctx, host)
 	if err != nil {
 		return err
 	}
 
+	if !registrationDone {
+		se.logger.Info("Checking if locally retrieved credentials are still valid...",
+			zap.String(collectorNameField, colCreds.Credentials.CollectorName),
+			zap.String(collectorIdField, colCreds.Credentials.CollectorId),
+		)
+
+		if err := se.validateCredenials(ctx, colCreds.Credentials); err != nil {
+			se.logger.Info("Locally stored credentials invalid. Trying to re-register...")
+			if err = se.credentialsStore.Delete(se.hashKey); err != nil {
+				return err
+			}
+
+			// Credentials might have ended up being invalid or the collector
+			// might have been removed in Sumo.
+			// Fall back to removing the credentials and recreating them.
+			colCreds, _, err = se.handleCredentials(ctx, host)
+			if err != nil {
+				return fmt.Errorf("registration failed: %w", err)
+			}
+			se.logger.Info("Registration finished successfully",
+				zap.String(collectorNameField, colCreds.Credentials.CollectorName),
+				zap.String(collectorIdField, colCreds.Credentials.CollectorId),
+			)
+		} else {
+			se.logger.Info("Local collector credentials all good, starting up the collector",
+				zap.String(collectorNameField, colCreds.Credentials.CollectorName),
+				zap.String(collectorIdField, colCreds.Credentials.CollectorId),
+			)
+		}
+	}
```

Comment:
This flow of logic is a bit confusing to me. As far as I understand it, what we want is:
1. Check if we have local credentials and if they're valid.
2. If not, register with the remote.
This code seems to call `handleCredentials` a second time if the first time fails, and it's not clear why the result should be different.

In general, I think we'd benefit from clearly separating acting on the remote from doing things locally. I'd probably split `getCredentials` into two separate functions and be explicit about the current state in this logic.

Reply:
Perhaps you're right with the fact that it's a bit twisted 🤔 

The result would be different because we've deleted the credentials stored locally (line 165). Do you have an idea how you would see this?

Reply:
At a high level, I think I'd like this to look something like:
```go
if se.conf.Clobber {
  colCreds, err := se.getRemoteCredentials()
  return colCreds, err
}
colCreds, exists := se.getLocalCredentials()
if exists && se.validateCredentials(colCreds) {
  return colCreds
}

colCreds, err := se.getRemoteCredentials()
return colCreds, err
```
just for getting the credentials. Then they can be stored locally higher up the call chain.

Reply:
Alrighty then but that looks like refactoring which would ideally land in a separate PR :)

Reply:
Fair enough, can you add a TODO here then, if you're not going to do the refactor immediately?

Reply:
Done